### DISCUSSION TITLE
Using trained model(NHITS) for predicting over multiple horizons

### DISCUSSION THREAD

**Discussion by shwva184** [Nixtla/neuralforecast#449] [Q&A]
If one needs to use a trained NHITS model (using Autonhits) for a horizon of 1 week ahead ( 168 hours ahead) for the upcoming month by just changing the training data as new data is available, how can we do it?

Assume model1 is the trained and tuned model on training data Y_train, I tried doing the following to obtain results for upcoming 4 weeks.

week 1: model1.predict(Y_train)
week 2: model1.predict(Y_train+data from week1)
... and so on

But the obtained predictions for all these week are the same in spite of changing the data in predict function.
Is there any other way of achieving this?

Many thanks in advance

**Comment by jimbow973** [comment]
Hi, i am stuck with the same problem. Have you found a solution?
thanks

**Comment by jmoralez** [comment]
Hey. Can you provide a reproducible example? The following works:

```python
import os

import pandas as pd
from neuralforecast import NeuralForecast
from neuralforecast.models import LSTM
from neuralforecast.utils import AirPassengersDF

os.environ['NIXTLA_ID_AS_COL'] = '1'

nf = NeuralForecast(
    freq='M',
    models=[LSTM(h=12, input_size=24, max_steps=1)]
)
nf.fit(AirPassengersDF)

# predict first time
preds = nf.predict()

# incorporate predictions into train df (these can be your real values instead)
new_df = pd.concat([AirPassengersDF, preds.rename(columns={'LSTM': 'y'})])

# predict again using this extended dataframe
new_preds = nf.predict(df=new_df)
print(pd.concat([preds, new_preds[['ds', 'LSTM']]], axis=1))  # these are all different
```

**Reply by mdommett** [reply]
hello 👋  

Thanks for providing this example! I have a follow up question for this example. 

From the example, in order to predict a horizon h+1 after h0 , you have appended h0 onto the training data. 

Is this always necessary? 

For example, if I have trained a model, and want to use that trained model for real-time forecasting the horizon every X minutes, must I always keep the training data in memory? ie in an NIHTS (or any neuralforecast model),  to predict any horizon after the training set, it is required to store the training data PLUS any previous data between the training data and the next horizon? 

Thanks for your help in clarifying!

**Reply by jmoralez** [reply]
Hey. You need to keep at least `input_size` samples in each serie, so that the model gets the context it needs to predict (24 in this case).

**Reply by mdommett** [reply]
thanks!

**Reply by mdommett** [reply]
Hello @jmoralez ,

Coming back with another question on this 😬 

I am using `NHITS` model , but it seems to not require the given `df` to be of length `input_size` (ie with the context). Here is a working example:

```
from neuralforecast.auto import NHITS
import os

import pandas as pd
from neuralforecast import NeuralForecast
from neuralforecast.models import LSTM
from neuralforecast.utils import AirPassengersDF

nfm = NeuralForecast(
    freq='M',
    models=[NHITS(h = 12,
                  max_steps=50,
                input_size = 24,
                scaler_type = 'robust'),      
                ]
)
nfm.fit(AirPassengersDF)

# predict first time
preds = nfm.predict()

# incorporate predictions into train df (these can be your real values instead)
new_df = pd.concat([AirPassengersDF, preds.rename(columns={'NHITS': 'y'})])

# predict again using this extended dataframe BUT only giving an input size of 1
new_preds = nfm.predict(df=new_df.tail(1))
print(pd.concat([preds, new_preds[['ds', 'NHITS']]], axis=1)) 
```

The `nfm.predict(df=new_df.tail(1))` I would expect to fail since the context is not of length 24, but it works. How does the `NHITS` model therefore know the context? 

Any help appreciated! Thanks

**Reply by jmoralez** [reply]
It uses 23 zeros and the 1 sample you gave it to predict, so it's most likely very bad. The requirement isn't strict in the sense that it won't produce an error, but you'll get the most accurate forecasts by providing the full input size.

**Reply by mdommett** [reply]
thanks 🙏
